shotover-proxy/src/transforms/protect/mod.rs

```diff
@@ -30,6 +33,161 @@ pub struct Protect {
+    }
+
+    /// get the list of protected columns for the specified table name.  Will return `None` if no columns
+    /// are defined for the table.
+    fn get_protected_columns(&self, table_name: &FQName) -> Option<&Vec<String>> {
+        // TODO replace "" with cached keyspace name
+        if let Some(tables) = self
+            .keyspace_table_columns
+            .get(table_name.extract_keyspace(""))
+        {
+            tables.get(&table_name.name)
+        } else {
+            None
+        }
+    }
+
+    /// extractes the protected object from the message value.  Resulting object is a Protected::Ciphertext
+    fn extract_protected(&self, value: &MessageValue) -> Result<Protected> {
+        match value {
+            MessageValue::Bytes(b) => Protect::decode(&b[..]),
+            MessageValue::Varchar(hex_value) => {
+                let byte_value = hex::decode(hex_value);
+                Protect::decode(&byte_value.unwrap())
+            }
+            _ => Err(anyhow!(
+                "Could not get bytes to decrypt - wrong value type {:?}",
+                value
+            )),
+        }
+    }
+
+    /// determines if columns in the CassandraStatement need to be encrypted and encrypts them.  Returns `true` if any columns were changed.
+    ///  * `statement` the statement to encrypt.
+    ///  * `columns` the column names to encrypt.
+    ///  * `key_source` the key manager with encryption keys.
+    ///  * `key_id` the key within the manager to use.
+    async fn encrypt_columns(&self, statement: &mut CassandraStatement) -> Result<bool> {
+        let mut data_changed = false;
+        if let Some(table_name) = cql_statement::get_table_name(statement) {
+            if let Some(columns) = self.get_protected_columns(table_name) {
+                match statement {
+                    CassandraStatement::Insert(insert) => {
+                        // get the indices of the inserted protected columns
+                        let indices: Vec<usize> = insert
+                            .columns
+                            .iter()
+                            .enumerate()
+                            .filter_map(|(i, col_name)| {
+                                if columns.contains(col_name) {
+                                    Some(i)
+                                } else {
+                                    None
+                                }
+                            })
+                            .collect();
+                        // if there are columns process them
+                        if !indices.is_empty() {
+                            match &mut insert.values {
+                                InsertValues::Values(value_operands) => {
+                                    for idx in indices {
+                                        let mut protected = Protected::Plaintext(
+                                            MessageValue::from(&value_operands[idx]),
```

Comment:
`value_operands[idx]` will panic if the user doesnt provide enough values.
Maybe something like this would work well:
```rust
if let Some(value)  = value_operands.get_mut(idx) {
    let mut protected = Protected::Plaintext(
       MessageValue::from(value),
    );
    protected = protected
      .protect(&self.key_source, &self.key_id)
      .await?;
    value = Operand::from(&protected);
    data_changed = true
}
```
(sorry for any compiler errors, I dont have access to a proper dev environment today)